File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/gui/EcoGUIFactory.java
package com.willfp.eco.internal.gui;

import com.willfp.eco.core.gui.GUIFactory;
import com.willfp.eco.core.gui.menu.Menu;
import com.willfp.eco.core.gui.slot.Slot;
import com.willfp.eco.internal.gui.menu.EcoMenuBuilder;
import com.willfp.eco.internal.gui.slot.EcoSlotBuilder;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.function.Function;

public class EcoGUIFactory implements GUIFactory {
    @Override
    public Slot.Builder createSlotBuilder(@NotNull final Function<Player, ItemStack> provider) {
        return new EcoSlotBuilder(provider);
    }

    @Override
    public Menu.Builder createMenuBuilder(final int rows) {
        return new EcoMenuBuilder(rows);
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/gui/EcoGUIFactory.kt
package com.willfp.eco.internal.gui

import com.willfp.eco.core.gui.GUIFactory
import com.willfp.eco.core.gui.menu.Menu
import com.willfp.eco.core.gui.menu.MenuBuilder
import com.willfp.eco.core.gui.slot.functional.SlotProvider
import com.willfp.eco.internal.gui.menu.EcoMenuBuilder
import com.willfp.eco.internal.gui.page.MergedStateMenu
import com.willfp.eco.internal.gui.slot.EcoSlotBuilder

object EcoGUIFactory : GUIFactory {
    override fun createSlotBuilder(provider: SlotProvider) =
        EcoSlotBuilder(provider)

    override fun createMenuBuilder(rows: Int): MenuBuilder =
        EcoMenuBuilder(rows)

    override fun blendMenuState(base: Menu, additional: Menu): Menu =
        MergedStateMenu(base, additional)
}
